package/MDAnalysis/analysis/hbonds/wbridge_analysis.py

```diff
@@ -0,0 +1,821 @@
+        try:
+            frames = np.arange(self.u.trajectory.n_frames)[self.traj_slice]
+        except:
+            logger.error("Problem reading trajectory or trajectory slice incompatible.")
+            logger.exception()
+            raise
+        verbose = _set_verbose(verbose=kwargs.get('verbose', None),
+                               quiet=kwargs.get('quiet', None),
+                               default=True)
+        pm = ProgressMeter(len(frames),
+                           format="WBridge frame {current_step:5d}: {step:5d}/{numsteps} [{percentage:5.1f}%]\r",
+                           verbose=verbose)
+
+        try:
+            self.u.trajectory.time
+            def _get_timestep():
+                return self.u.trajectory.time
+            logger.debug("WBridge analysis is recording time step")
+        except NotImplementedError:
+            # chained reader or xyz(?) cannot do time yet
+            def _get_timestep():
+                return self.u.trajectory.frame
+            logger.warning("WBridge analysis is recording frame number instead of time step")
+
+        logger.info("Starting analysis (frame index start=%d stop=%d, step=%d)",
+                    (self.traj_slice.start or 0),
+                    (self.traj_slice.stop or self.u.trajectory.n_frames), self.traj_slice.step or 1)
+
+        for progress, ts in enumerate(self.u.trajectory[self.traj_slice]):
+            # all bonds for this timestep
+
+            # dict of tuples (atom.index, atom.index) for quick check if
+            # we already have the bond (to avoid duplicates)
+
+            frame = ts.frame
+            timestep = _get_timestep()
+            self.timesteps.append(timestep)
+
+            pm.echo(progress, current_step=frame)
+            self.logger_debug("Analyzing frame %(frame)d, timestep %(timestep)f ps", vars())
+            if self.update_selection1:
+                self._update_selection_1()
+            if self.update_selection2:
+                self._update_selection_2()
+            if self.update_water_selection:
+                self._update_water_selection()
+
+            s1_frame_results_dict = {}
+            if self.selection1_type in ('donor', 'both') and self._water_acceptors:
+                self.logger_debug("Selection 1 Donors <-> Water Acceptors")
+                ns_acceptors = AtomNeighborSearch(self._water_acceptors)
+                for i, donor_h_set in self._s1_donors_h.items():
+                    d = self._s1_donors[i]
+                    for h in donor_h_set:
+                        res = ns_acceptors.search(h, self.distance)
+                        for a in res:
+                            angle = self.calc_angle(d, h, a)
+                            donor_atom = h if self.distance_type != 'heavy' else d
+                            dist = self.calc_eucl_distance(donor_atom, a)
+                            if angle >= self.angle and dist <= self.distance:
```

Comment:
As an optimisation you could check either the `dist` or `angle` first, and then only check the other if the first one passed  (I'd guess distance criteria is harder to meet, so check that first).

With Python's if syntax, the components are evaluated lazily around the `and`... so the following would do the above optimisation:

```python
if self.calc_eucl_distance(donor_atom, a) <= self.distance and self.calc_angle(d, h, a) >= self.angle:

```

ie `self.calc_angle` is only called if the first check passed

Reply:
@richardjgowers Thank you for the advice. The distance check is now performed before angle calculation.